python/private/bzlmod/pypi_archive.bzl

```diff
@@ -0,0 +1,137 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"TODO"
+
+load("//python:versions.bzl", "WINDOWS_NAME")
+load("//python/private:auth.bzl", "get_auth")
+load("//python/private:bzlmod_enabled.bzl", "BZLMOD_ENABLED")
+load("//python/private:patch_whl.bzl", "patch_whl")
+load("//python/private:toolchains_repo.bzl", "get_host_os_arch")
+
+_HTTP_FILE_DOC = """See documentation for the attribute with the same name
+in [http_file docs](https://bazel.build/rules/lib/repo/http#http_file document)."""
+
+def _impl(rctx):
+    prefix, _, _ = rctx.attr.name.rpartition("_")
+    prefix, _, _ = prefix.rpartition("_")
+
+    _, _, filename = rctx.attr.urls[0].rpartition("/")
+    filename = filename.strip()
+
+    urls = rctx.attr.urls
+    auth = get_auth(rctx, urls)
+
+    result = rctx.download(
+        url = urls,
+        output = filename,
+        sha256 = rctx.attr.sha256,
+        auth = auth,
+        canonical_id = rctx.attr.canonical_id,
+        integrity = rctx.attr.integrity,
+    )
+    if not result.success:
+        fail(result)
+
+    whl_path = rctx.path(filename)
+
+    if rctx.attr.patches:
+        whl_path = patch_whl(
+            rctx,
+            python_interpreter = _resolve_python_interpreter(rctx),
+            whl_path = whl_path,
+            patches = rctx.attr.patches,
+            quiet = rctx.attr.quiet,
+        )
+
+    rctx.symlink(whl_path, "file")
+
+    rctx.file(
+        "BUILD.bazel",
+        """\
+filegroup(
+    name="file",
```

Comment:
This target name conflicts with the symlink created above. I think targets will hide the source file, but its not a good practice.